### PR TITLE
pulselistener: Trigger jobs on try on new phabricator diffs

### PR DESCRIPTION
**Pull request by La0**
The goal is to trigger a try job when a new phabricator message is received by the Phabricator ~~poller~~ listener.

A new async process is added to pulselistener:

- it manages a unique local clone of **Mozilla unified**
- it has an asyncio queue, and continously reads it
- the same queue is used by the phabricator process to push new diffs
- when a new diff is read, the stack of patches is retrieved, applied and pushed to try through ssh

### Misc.

- The code from `staticanalysis/bot` that retrieves a phabricator stack of patches is moved in `lib/cli_common` and now shared by `staticanalysis/bot` and `pulselistener`
- `pulselistener` uses the common `mercurial.nix` with all needed options (instead of building its own)
- `pulselistener` now needs `openssh`
- `pulselistener` now has an optiona argument `--cache-root=/path/to/cache`; it's used to clone the repo. It defaults to `/tmp/pulselistener`

### Taskcluster secrets updates

The phabricator configuration is moved from the hook configuration to the main secret:

```
PHABRICATOR:
  url: https://phabricator...
  token: XXX
```

An ssh configuration is added to be able to push on try:

```
MERCURIAL_REMOTE:
  ssh_user: xxx@mozilla.com
  ssh_key: superlongsshkey...
```

An optional `actions` list is added to the phabricator hook configuration. It supports two values:

- `taskcluster` to activate direct taskcluster task creation (current mode)
- `try` to activate push on try

By default, only `taskcluster` is enabled

### CODE REVIEW

File: lib/cli_common/cli_common/phabricator.py [hunk 1]
```diff
@@ -26,6 +28,17 @@ def revision_exists_on_central(revision):
     return resp.ok
 
 
+def revision_available(repo, revision):
+    '''
+    Check a revision is available on a Mercurial repo
```
Comment (marco-c) [inline lib/cli_common/cli_common/phabricator.py]: Nit: Check if

File: lib/cli_common/cli_common/phabricator.py [hunk 2]
```diff
@@ -324,3 +337,65 @@ def request(self, path, **payload):
         # Outputs result
         assert 'result' in data
         return data['result']
+
+    def load_patches_stack(self, repo, diff, default_revision='central'):
```
Comment (marco-c) [inline lib/cli_common/cli_common/phabricator.py]: You just moved this from static analysis, right? (so I don't have to review it)

File: src/pulselistener/pulselistener/listener.py [hunk 3]
```diff
@@ -46,9 +47,12 @@ def __init__(self, configuration):
         assert len(self.repos) > 0, 'No repositories enabled'
         logger.info('Enabled Phabricator repositories', repos=[r['fields']['name'] for r in self.repos.values()])
 
+        # Get actions to do on new diff
+        self.actions = configuration.get('actions', [ACTION_TASKCLUSTER, ])
+        logger.info('Enabled actions', actions=self.actions)
+
         # Start by getting top id
         diffs = self.api.search_diffs(limit=1)
-        assert len(diffs) == 1
```
Comment (marco-c) [inline src/pulselistener/pulselistener/listener.py]: Why did you remove this assert?

File: src/pulselistener/pulselistener/mercurial.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+from cli_common.log import get_logger
+from cli_common.mercurial import robust_checkout
+from pulselistener.config import REPO_TRY
+
+logger = get_logger(__name__)
+
+
+class MercurialWorker(object):
+    '''
+    Mercurial worker maintaining a local clone of Mozilla-Central
```
Comment (marco-c) [inline src/pulselistener/pulselistener/mercurial.py]: Nit: mozilla-central with lower case characters :P

**Comment by La0** [comment]
Yes!

File: src/pulselistener/pulselistener/mercurial.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        self.clean()
+
+        # Get the stack of patches
+        patches = self.phabricator_api.load_patches_stack(self.repo, diff, default_revision='central')
+        assert len(patches) > 0, 'No patches to apply'
+
+        # Get current revision
+        base = self.repo.tip()
+
+        # Apply the patches, without commiting
```
Comment (marco-c) [inline src/pulselistener/pulselistener/mercurial.py]: Why not applying the patches one by one, committing each one?

File: src/pulselistener/pulselistener/mercurial.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    async def run(self):
+        # Start by updating the repo
+        logger.info('Checking out tip', repo=self.repo_url)
+        self.repo = robust_checkout(self.repo_url, self.repo_dir)
+        self.repo.setcbout(lambda msg: logger.info('Mercurial', stdout=msg))
+        self.repo.setcberr(lambda msg: logger.info('Mercurial', stderr=msg))
+        logger.info('Initial clone finished')
+
+        # Wait for phabricator diffs to apply
+        while True:
```
Comment (marco-c) [inline src/pulselistener/pulselistener/mercurial.py]: We should pull the changes from mozilla-unified, otherwise our local clone will get out of date.

File: src/pulselistener/pulselistener/mercurial.py (hunk 5 above)
Comment (La0) [reply src/pulselistener/pulselistener/mercurial.py]: I'm now commiting each changes to avoid merge conflicts, i just need to update the comment

File: lib/cli_common/cli_common/phabricator.py (hunk 2 above)
Comment (La0) [reply lib/cli_common/cli_common/phabricator.py]: Yeap :+1:

File: src/pulselistener/pulselistener/listener.py (hunk 3 above)
Comment (La0) [reply src/pulselistener/pulselistener/listener.py]: no idea why

File: src/pulselistener/pulselistener/mercurial.py (hunk 6 above)
Comment (La0) [reply src/pulselistener/pulselistener/mercurial.py]: Good idea, i was thinking of updating every X minutes in a later commit, but that may be overkill for now

Comment (marco-c) [reply src/pulselistener/pulselistener/mercurial.py]: We could also update once every X revisions, which is an approximation of once every X minutes and might be easier to implement.

Comment (La0) [reply src/pulselistener/pulselistener/mercurial.py]: I simply added the update on each diff, it's pretty fast as it keeps getting updated. We'll see how it behaves once on Heroku

File: src/pulselistener/pulselistener/mercurial.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        # Apply the patches and commit them one by one
+        for diff_phid, patch in patches:
+            logger.info('Applying patch', phid=diff_phid)
+            self.repo.import_(
+                patches=io.BytesIO(patch.encode('utf-8')),
+                message='Patch {}'.format(diff_phid),
+                user='pulselistener',
+            )
+
+        # Rebase multiple patches into a single commit
```
Comment (marco-c) [inline src/pulselistener/pulselistener/mercurial.py]: Squashing the patches into a single commit shouldn't be needed

Comment (La0) [reply src/pulselistener/pulselistener/mercurial.py]: Agreed, it works without squash: https://treeherder.mozilla.org/#/jobs?repo=try&revision=c7cfff789d14b9ba7b25ad329df72639af03ba65&selectedJob=226153182

File: src/pulselistener/pulselistener/mercurial.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+        # Start by cleaning the repo
+        self.clean()
+
+        # Get the stack of patches
+        patches = self.phabricator_api.load_patches_stack(self.repo, diff, default_revision='central')
+        assert len(patches) > 0, 'No patches to apply'
+
+        # Get current revision
+        base = self.repo.tip()
```
Comment (marco-c) [inline src/pulselistener/pulselistener/mercurial.py]: We should use the base of the stack (which is the changeset we are at right now, since load_patches_stack updates the repo to the base of the stack) rather than the tip of the repo (the tip is the latest commit, no?).

File: src/pulselistener/tests/test_mercurial.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
```
Comment (marco-c) [inline src/pulselistener/tests/test_mercurial.py]: It'd be nice to also test the case where a stack of revisions applies on top of a base revision which is not the latest in central, but we can also add this test later on.

Comment (La0) [reply src/pulselistener/tests/test_mercurial.py]: I added a unit test for this case

File: src/pulselistener/pulselistener/mercurial.py (hunk 8 above)
Comment (La0) [reply src/pulselistener/pulselistener/mercurial.py]: Now `load_stack_patches()` gives the mercurial base revision it updated the repo to + the patches
